ComWrappers: Add support for ICustomQueryInterface

AaronRobinsonMSFT opened the pull request:
If the managed object wrapper doesn't know about the IID but implements [`ICustomQueryInterface`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.icustomqueryinterface.getinterface), ask the implementation to find the interface.

Fixes https://github.com/dotnet/runtime/issues/34717

/cc @jkoritzinsky @elinor-fung @jkotas @davidwrighton

AaronRobinsonMSFT committed:
Add support for falling back to ICustomQueryInterface if the managed object wrapper
 doesn't know about the IID but implements the interface.

AaronRobinsonMSFT committed:
Applying some feedback.

jkotas commented on src/coreclr/src/interop/comwrappers.cpp:
```diff
@@ -584,7 +584,22 @@ HRESULT ManagedObjectWrapper::QueryInterface(
     // Find target interface
     *ppvObject = As(riid);
     if (*ppvObject == nullptr)
```
The `ICustomQueryInterface` contract for regular objects is "first chance". `ICustomQueryInterface` gets called before the default implementation kicks in. This one is one is "last change" - it gets called after the default implementation kicks in.

We can talk about it tomorrow...

AaronRobinsonMSFT replied:
Yep. For `ComWrappers`, that isn't really possible due to the GC issues. I also don't think it is the right semantics either. For the built-in case, we provide several default implementations that user's don't even have a say over, but in `ComWrappers` it is entirely driven by the author last seems right to me. I do agree the semantics are inverted here and... concerning given how we originally scoped/designed the `ComWrappers` feature.

AaronRobinsonMSFT committed:
Follow ICustomQueryInterface contract semantics.

AaronRobinsonMSFT committed:
Handle the case a GC thread is being used to QI a wrapper.

AaronRobinsonMSFT committed:
The QI on the managed object wrapper is now callable from within the runtime.
This means we can no longer assume we are in preemptive mode during the QI.
Update the ICustomQueryInterface dispatch to be okay with ANY mode.

Tests needed to be updated with the ICustomQueryInterface support.

AaronRobinsonMSFT commented:
These changes are limited to the Windows platforms so merging.
